### PR TITLE
Security Update & Role-Based Screens

### PR DESCRIPTION
**Pull request by aptkingston**
## Description
This PR adds a better implementation of role based security, linking sets of permissions with roles. Roles can be managed and assigned to users in the users table UI in the backend.

* It's now possible to have conditional client app rendering, based on roles. Each screen has a role setting, which specifies the minimum required role to access the page.
* Viewing screens has changed slightly, and is now filtered by role instead. This is required as multiple screens could have the same route, but be different depending on the user's role.
* It's possible to create multiple screens with the same route but different access level. This means you could make a homepage which shows basic info for a normal user, and rich graphs for an admin user instead.

There are a lot of other improvements around selecting and displaying screens which should hopefully fix a few of the intermittent crashes. Performance should be snappier too, due to ripping out a lot of legacy code surrounding generating CSS.

## Screenshots
Roles are managed from the users table
roles

Roles can be assigned to users
role assignment

Screens can be assigned a minimum access role
settings

Screens are filtered based on access role
selector

## App example
Here's an example of an app with a different home screen depending on the user's role.

Logging in as an admin
admin

Logging in as a basic user
basic

### CODE REVIEW

**Comment by joebudi** [comment]
@aptkingston Should we change 'Add new row' to 'Add new user'?

**Comment by aptkingston** [comment]
@joebudi yep definitely. I've renamed the rest of the instances of "row" to "user" so I'll do that one too.

File: packages/builder/src/components/backend/DataTable/Table.svelte [hunk 1]
```diff
@@ -57,23 +73,34 @@
...
         },
       ]
     }
 
-    Object.keys(schema || {}).forEach((key, idx) => {
+    const canEditColumn = key => {
+      if (!allowEditing) {
+        return false
+      }
+      return !(isUsersTable && ["email", "roleId"].indexOf(key) !== -1)
```
Comment (shogunpurple) [inline packages/builder/src/components/backend/DataTable/Table.svelte:85]: NAB: `includes` works here too

**Review by shogunpurple — APPROVED** [review]
Great job - this feature is awesome. Solid bit of refactoring in there as well. 🚀

**Comment by kevmodrome** [comment]
Would it be nice to have a write only permission? Would allow users to enter data but not read anything

**Comment by kevmodrome** [comment]
Found this odd thing playing with adding users from the frontend: https://cln.sh/4YfqA6

Side note, do we not want the possibility to also specify passwords if we're adding new users like this?

**Comment by kevmodrome** [comment]
I think another key feature for this would be to also have different layouts for different roles, it's a bit strange to have links to pages you can't access (if that makes sense)

**Review by kevmodrome — APPROVED** [review]
Looks good! 👍 🏅

**Comment by joebudi** [comment]
I've pulled down the PR and tested it. Comments below:

- When you create or edit a layout, and click the Screens tab, it always brings you back to the basic screen. This was annoying as I was working on the public screens.
https://www.loom.com/share/a1c649e60c164e85b7d4f2b73240bf65

- I like the separation of screens by user role. This makes sense and will reduce user error in the future.

- When I create a new public screen, it redirects me to the login page when I try to hit that URL. If I change the login URL from "/" to "/login" and change my new screen to "/" I can access my new screen and login. But! On that new screen, data components will not display (i tried a chart, table, card). Static components do display (I tried an image and text).

- I can't log in to my apps.
https://www.loom.com/share/3b96c9fed9564e9e9f50ae11a730ca4c

**Comment by aptkingston** [comment]
@kevmodrome 
I wish I could reply directly to the different comments!

> Would it be nice to have a write only permission? Would allow users to enter data but not read anything

Personally I don't see a use case for write-only (if you write it, should you not be able to read it?), but we can add it if we need to.

>Found this odd thing playing with adding users from the frontend: https://cln.sh/4YfqA6. Side note, do we not want the possibility to also specify passwords if we're adding new users like this?

User's shouldn't really be linked up to a new row or row detail component I don't think. Roles aren't an actual table (which is why it looks weird with ID's there) and passwords aren't in the user schema either. Managing users in the front end isn't specific to this PR though, like I'm sure that's been possible since the users-as-a-table PR. I'd be in favour or removing the users table as a data source for the frontend entirely. If we wanted user management we would need custom components for users, the same way there are now custom backend components in use for managing them within the builder.

> I think another key feature for this would be to also have different layouts for different roles, it's a bit strange to have links to pages you can't access (if that makes sense)

Yea I thought about this too. The only real solutions are different layouts for each role, or adding in conditional UI to components, so that each component will only be rendered if the user is a certain role. Conditional components basically makes unique screens per role redundant though... And layouts per role  adds even more complexity. I'm really not sure what to do about it.

You also don't need to have routes visible that you can't access though. Most of the time there will be an equivalent screen for a normal user, just with different content. Worse case, if it's an "admin only" screen with no equivalent, you could put a link to it in your admin-specific home page rather than the nav bar. But I realise that's not ideal either.

I'm nearly leaning down the route of conditional components (which could be a screen too - screens are components). Means the client app would just be in charge of choosing what to render but it would get the full definitions. Would love other opinions on this thought.

**Comment by aptkingston** [comment]
@joebudi 

>When you create or edit a layout, and click the Screens tab, it always brings you back to the basic screen. This was annoying as I was working on the public screens.

This is because we don't have a separate state entry for "selected screen" and "selected layout" - they share the same variable. So when you go to layouts, it "forgets" what screen was selected, and when you go back to screens it'll just pick the first one. We would need to refactor to store both separately. This has always been the case, it's just with the role-filtering it's more obvious, as the screen you had previously selected might not be visible at all - whereas previously the screen you wanted would be visible but just not selected.

> When I create a new public screen, it redirects me to the login page when I try to hit that URL. If I change the login URL from "/" to "/login" and change my new screen to "/" I can access my new screen and login. But! On that new screen, data components will not display (i tried a chart, table, card). Static components do display (I tried an image and text).

I'll need to play around and see if I can replicate this. Data components won't work on a public screen (unless you're logged in). I think the API will reject any requests to read data if you aren't authenticated. This will change whenever we implement role based access for tables, because then you could specify a "public" table, which would allow anyone to read it even if they aren't logged in.

> I can't log in to my apps.

You can and you actually are logged in. You're just still on the login page because you changed the login URL to `/login`. Normally apps will have both login and home on the `/` route, which means the server will only return the home page once they've logged in. In your case, because `/login` is a unique URL it's still being returned. There is an easy fix for this - when you log in I'll push you to `/`. Currently it just reloads the page which is the problem, and why you're still on the login page.

**Comment by kevmodrome** [comment]
> Personally I don't see a use case for write-only (if you write it, should you not be able to read it?), but we can add it if we need to.
> 

Yeah it's probably a very niche use-case. I'm imagining some odd compliance use-case, maybe medical, where the user is not under any circumstance allowed to read patient data but is allowed to add to it. Maybe? Does that make sense?

> User's shouldn't really be linked up to a new row or row detail component I don't think. Roles aren't an actual table (which is why it looks weird with ID's there) and passwords aren't in the user schema either. Managing users in the front end isn't specific to this PR though, like I'm sure that's been possible since the users-as-a-table PR. I'd be in favour or removing the users table as a data source for the frontend entirely. If we wanted user management we would need custom components for users, the same way there are now custom backend components in use for managing them within the builder.
> 

Yes, this makes sense. I think it should be removed if you aren't able to use it as a regular table, at least some kind of warning should show up somewhere, imo!

> > I think another key feature for this would be to also have different layouts for different roles, it's a bit strange to have links to pages you can't access (if that makes sense)
> 
> Yea I thought about this too. The only real solutions are different layouts for each role, or adding in conditional UI to components, so that each component will only be rendered if the user is a certain role. Conditional components basically makes unique screens per role redundant though... And layouts per role adds even more complexity. I'm really not sure what to do about it.
> 
> You also don't need to have routes visible that you can't access though. Most of the time there will be an equivalent screen for a normal user, just with different content. Worse case, if it's an "admin only" screen with no equivalent, you could put a link to it in your admin-specific home page rather than the nav bar. But I realise that's not ideal either.
> 
> I'm nearly leaning down the route of conditional components (which could be a screen too - screens are components). Means the client app would just be in charge of choosing what to render but it would get the full definitions. Would love other opinions on this thought.

I think I agree with you here. Conditional components make more sense from a developer experience as well, I think.

**Comment by aptkingston** [comment]
As part of making a change to ensure that toggling between layouts and pages keeps the previously selected screen visible (by ensuring the selected role is preserved), now whenever you change the access role for a screen it won't automatically select that role. Hard to explain, easier to see in practice. But basically there is a UX tradeoff here in terms of what is better. I might just go down the route of refactoring the frontend store more, to support both a selected screen and selected layout property.

**Review by mike12345567 — APPROVED** [review]
LGTM! Can't wait to have this in.

**Pull request merged**